system/codesigner/gpgcodesigner.cpp

```diff
@@ -0,0 +1,278 @@
+    return unsignedText.append(tail-head, head);
+}
+
+const char* skipn(const char* str, char c, int n)
+{
+    for (int i = 0; i < n && str && *str; i++)
+    {
+        str = strchr(str, c);
+        if (!str)
+            break;
+        str++;
+    }
+    return str;
+}
+
+StringArray &GpgCodeSigner::getUserIds(StringArray &userIds, StringBuffer &errmsg)
+{
+    initGpg();
+    StringBuffer output("\n");
+    errmsg.clear();
+    int ret = runExternalCommand(output, errmsg, "gpg --list-secret-keys --with-colon", nullptr);
+    if (ret != 0)
+        throw MakeStringException(-1, "Error running gpg: %s", errmsg.str());
+
+    const char* START = "\nuid:";
+    if (isGpgV1)
+        START = "\nsec:";
+    int startlen = strlen(START);
+    const int SKIP = 8;
+    const char* line = output.str();
+    StringArray uids;
+    while (line && *line)
+    {
+        line = strstr(line, START);
+        if (!line)
+            break;
+        line += startlen;
+        line = skipn(line, ':', SKIP);
+        if (!*line)
+            break;
+        const char* uid_s = line;
+        while (*line != '\0' && *line != ':')
+            line++;
+        if (line > uid_s)
+        {
+            StringBuffer uid(line - uid_s, uid_s);
+            uid.trim();
+            if (uid.length() > 0)
+                uids.append(uid.str());
+        }
+    }
+    uids.sortAscii(false);
+    const char* current = "";
+    for (unsigned i = 0; i < uids.length(); i++)
+    {
+        if (strcmp(uids.item(i), current) != 0)
+        {
+            current = uids.item(i);
+            userIds.append(current);
+        }
+    }
+
+    return userIds;
+}
+
+bool GpgCodeSigner::getKeyGrip(const char *userId, StringBuffer &keygrip)
+{
+    initGpg();
+    keygrip.clear();
+
+    StringBuffer cmd;
+    if (isGpgV1)
+        cmd.appendf("gpg --list-secret-keys \"=%s\"", userId); // = means exact match
+    else
+        cmd.appendf("gpg --list-secret-keys --with-keygrip \"=%s\"", userId); // = means exact match
+
+    StringBuffer output, errmsg;
+    int ret = runExternalCommand(output, errmsg, cmd.str(), nullptr);
+    if (ret != 0 || strstr(output.str(), userId) == nullptr)
+        return false;
+    auto kgptr = strstr(output.str(), "Keygrip = ");
+    if (kgptr)
+        keygrip.append(40, kgptr+10);
+    else
+        return false;
+    return true;
+}
+
+void GpgCodeSigner::clearPassphrase(const char* key)
+{
+    initGpg();
+    StringBuffer output, errmsg;
+    VStringBuffer cmd("gpg-connect-agent \"clear_passphrase --mode=normal %s\" /bye", key);
+    runExternalCommand(output, errmsg, cmd.str(), nullptr);
+}
+
+StringBuffer & GpgCodeSigner::sanitizeUserId(const char * userId, StringBuffer & sanitizedUserId)
+{
+    sanitizedUserId.clear();
+    sanitizedUserId.ensureCapacity(strlen(userId));
+    const char *p = userId;
+    while (*p && *p==' ') ++p;
+    while (*p)
+    {
+        if (isalnum(*p) || *p==' ' || *p=='.' || *p=='@' || *p =='<' || *p=='>')
```

Comment:
This seems to be overly restrictive, chars like -, _ should be accepted. Would it be better to be exclusive instead of inclusive, and escape as necessary?

Reply:
I see your point.  I see in the original code, you simply rejected any user id containing a quote. That seems like a decent solution: we could reject all user ids that have a quote character, control character or too long (say longer than 2000 characters)?

Reply:
Yes that should work.